ssl/quic/quic_channel.c

```diff
@@ -0,0 +1,1661 @@
+
+    /*
+     * After we move to a later EL we must not allow our peer to send any new
+     * bytes in the crypto stream on a previous EL. Retransmissions of old bytes
+     * are allowed.
+     *
+     * In practice we will only move to a new EL when we have consumed all bytes
+     * which should be sent on the crypto stream at a previous EL. For example,
+     * the Handshake EL should not be provisioned until we have completely
+     * consumed a TLS 1.3 ServerHello. Thus when we provision an EL the output
+     * of ossl_quic_rstream_available() should be 0 for all lower ELs. Thus if a
+     * given EL is available we simply ensure we have not received any further
+     * bytes at a lower EL.
+     */
+    for (i = QUIC_ENC_LEVEL_INITIAL; i < ch->rx_enc_level; ++i)
+        if (i != QUIC_ENC_LEVEL_0RTT &&
+            !crypto_ensure_empty(ch->crypto_recv[ossl_quic_enc_level_to_pn_space(i)])) {
+            /* Protocol violation (RFC 9001 s. 4.1.3) */
+            ossl_quic_channel_raise_protocol_error(ch, QUIC_ERR_PROTOCOL_VIOLATION,
+                                                   OSSL_QUIC_FRAME_TYPE_CRYPTO,
+                                                   "crypto stream data in wrong EL");
+            return 0;
+        }
+
+    rstream = ch->crypto_recv[ossl_quic_enc_level_to_pn_space(ch->rx_enc_level)];
+    if (rstream == NULL)
+        return 0;
+
+    return ossl_quic_rstream_read(rstream, buf, buf_len, bytes_read,
+                                  &is_fin);
+}
+
+static int ch_on_handshake_yield_secret(uint32_t enc_level, int direction,
+                                        uint32_t suite_id, EVP_MD *md,
+                                        const unsigned char *secret,
+                                        size_t secret_len,
+                                        void *arg)
+{
+    QUIC_CHANNEL *ch = arg;
+    uint32_t i;
+
+    if (enc_level < QUIC_ENC_LEVEL_HANDSHAKE || enc_level >= QUIC_ENC_LEVEL_NUM)
+        /* Invalid EL. */
+        return 0;
+
+
+    if (direction) {
+        /* TX */
+        if (enc_level <= ch->tx_enc_level)
+            /*
+            * Does not make sense for us to try and provision an EL we have already
+            * attained.
+            */
+            return 0;
+
+        if (!ossl_qtx_provide_secret(ch->qtx, enc_level,
+                                     suite_id, md,
+                                     secret, secret_len))
+            return 0;
+
+        ch->tx_enc_level = enc_level;
+    } else {
+        /* RX */
+        if (enc_level <= ch->rx_enc_level)
+            /*
+            * Does not make sense for us to try and provision an EL we have already
+            * attained.
+            */
+            return 0;
+
+        /*
+        * Ensure all crypto streams for previous ELs are now empty of available
+        * data.
+        */
+        for (i = QUIC_ENC_LEVEL_INITIAL; i < enc_level; ++i)
+            if (!crypto_ensure_empty(ch->crypto_recv[i])) {
```

Comment:
Should this actually be calling `ossl_quic_enc_level_to_pn_space` here?

Reply:
Yes, good catch.